### PR TITLE
Merge validate httproute

### PR DESCRIPTION
**Pull request by hzxuzhonghu**

### CODE REVIEW

File: pkg/config/validation/virtualservice.go [hunk 1]
```diff
@@ -19,68 +19,60 @@ import (
...
+	routeType := getHTTPRouteType(http, delegate)
 	// check for conflicts
-	if http.Redirect != nil {
-		if http.Fault != nil {
-			errs = appendErrors(errs, errors.New("HTTP route cannot contain both fault and redirect"))
-		}
+	errs = validateHTTPRouteConflict(http, routeType)
+
+	// check http route match requests
+	errs = appendErrors(errs, validateHTTPRouteMatchRequest(http, routeType))
```
Comment (hzxuzhonghu) [inline pkg/config/validation/virtualservice.go:55]: There are some diffs in http match requests

File: pkg/config/validation/virtualservice.go [hunk 2]
```diff
@@ -19,68 +19,60 @@ import (
...
-// TODO(@hzxuzhonghu): deduplicate with validateHTTPRoute
-func validateChainingHTTPRoute(http *networking.HTTPRoute) (errs error) {
+func validateHTTPRoute(http *networking.HTTPRoute, delegate bool) (errs error) {
+	routeType := getHTTPRouteType(http, delegate)
 	// check for conflicts
-	if http.Redirect != nil {
-		if http.Fault != nil {
-			errs = appendErrors(errs, errors.New("HTTP route cannot contain both fault and redirect"))
-		}
+	errs = validateHTTPRouteConflict(http, routeType)
```
Comment (hzxuzhonghu) [inline pkg/config/validation/virtualservice.go:52]: Also some diffs here

**Comment by hzxuzhonghu** [comment]
@ramaraochavali

**Review by ramaraochavali — APPROVED** [review]
Looks good. Once Minor question

File: pkg/config/validation/virtualservice.go [hunk 3]
```diff
@@ -19,68 +19,60 @@ import (
...
+func getHTTPRouteType(http *networking.HTTPRoute, isDelegate bool) HTTPRouteType {
+	if isDelegate {
+		return DelegateRoute
 	}
-	// check for conflicts
-	if http.Redirect != nil {
-		if len(http.Route) > 0 {
-			errs = appendErrors(errs, errors.New("HTTP route cannot contain both route and redirect"))
+
+	if features.EnableVirtualServiceDelegate {
```
[the remark this answers is not in the document]
Comment (hzxuzhonghu) [reply pkg/config/validation/virtualservice.go:39]: It is treated as independent one in the normal flow. I will test it if this rare case could cause something wrong

**Pull request merged**